<a href="https://colab.research.google.com/github/thomreid11/DAS7000_PORT1/blob/main/DAS7000_port1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Thomas Reid - DAS7000 - PORT1**

Exploratory Data Analysis (EDA) on Manhattan Property Dataset investigating property types and prices within Manhattan, New York

Link to data set: https://drive.google.com/file/d/1XRVnZ2Z_XkD1dDmF1uYBdjjF2WVAzg0g/view?usp=drive_link

#### **Importing necessary libraries**

In [37]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
import seaborn as sns

#### **Mounting Google Drive to Google Collab and importing dataset**

In [38]:
# Linking Google colab with Google Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [39]:
#Importing dataset path from recently lniked google drive
drive_path = '/content/drive/MyDrive/Manhattan property dataset csv.csv'

In [40]:
# Converting csv to df
df = pd.read_csv(drive_path)

In [41]:
# Printing the first 5 rows of data
df.head()

BOROUGH   NEIGHBORHOOD         BUILDING CLASS CATEGORY  \
0        1  ALPHABET CITY         01 ONE FAMILY DWELLINGS   
1        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS   
2        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS   
3        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS   
4        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS   

  TAX CLASS AT PRESENT  BLOCK  LOT  EASEMENT BUILDING CLASS AT PRESENT  \
0                    1    376   43       NaN                        S1   
1                   2B    372   19       NaN                        C7   
2                    2    396   26       NaN                        C4   
3                    2    396   27       NaN                        C4   
4                    2    400   11       NaN                        C4   

                  ADDRESS APARTMENT NUMBER  ...  RESIDENTIAL UNITS  \
0       743 EAST 6 STREET              NaN  ...                1.0   
1   292 EAST 3 STREET, 4B              NaN  ...                9.0   
2      638 EAST 14 STREET              NaN  ...               20.0   
3      640 EAST 14 STREET              NaN  ...               20.0   
4  510 EAST 5TH STREET, 9              NaN  ...               18.0   

   COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET GROSS SQUARE FEET  \
0               1.0          2.0             2,090             3,680   
1               1.0         10.0             2,401             6,920   
2               0.0         20.0             2,779            10,740   
3               0.0         20.0             2,831            10,900   
4               0.0         18.0             2,404             6,875   

  YEAR BUILT  TAX CLASS AT TIME OF SALE  BUILDING CLASS AT TIME OF SALE  \
0     1940.0                          1                              S1   
1     1920.0                          2                              C7   
2     1920.0                          2                              C4   
3     1920.0                          2                              C4   
4     1900.0                          2                              C4   

   SALE PRICE    SALE DATE  
0            0  01/23/2025  
1            0  02/07/2025  
2            0  05/01/2025  
3            0  05/01/2025  
4      280,000  04/29/2025  

[5 rows x 21 columns]

In [42]:
# Looking at the general dataset information; such as data type and counts of each variable
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18491 entries, 0 to 18490
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   BOROUGH                         18491 non-null  int64  
 1   NEIGHBORHOOD                    18491 non-null  object 
 2   BUILDING CLASS CATEGORY         18491 non-null  object 
 3   TAX CLASS AT PRESENT            18491 non-null  object 
 4   BLOCK                           18491 non-null  int64  
 5   LOT                             18491 non-null  int64  
 6   EASEMENT                        0 non-null      float64
 7   BUILDING CLASS AT PRESENT       18491 non-null  object 
 8   ADDRESS                         18491 non-null  object 
 9   APARTMENT NUMBER                8634 non-null   object 
 10  ZIP CODE                        18491 non-null  int64  
 11  RESIDENTIAL UNITS               9366 non-null   float64
 12  COMMERCIAL UNITS                

#### **Data Preprocessing**

In [43]:
# Making columns neat and uniform
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_').str.title()

In [44]:
# Correcting the data types
# Changing strings to floats and removing the commas
df['Land_Square_Feet'] = df['Land_Square_Feet'].str.replace(',', '', regex=False).astype(float)
df['Gross_Square_Feet'] = df['Gross_Square_Feet'].str.replace(',', '', regex=False).astype(float)
df['Sale_Price'] = df['Sale_Price'].str.replace(',', '', regex=False).astype(int)

# Changing the remaining data types
df['Sale_Date'] = pd.to_datetime(df['Sale_Date'])
df["Tax_Class_At_Time_Of_Sale"] = df["Tax_Class_At_Time_Of_Sale"].astype("string")
df["Tax_Class_At_Present"] = df["Tax_Class_At_Present"].astype("string")


# Changing 3 columns in one go for efficiency
unit_cols = ["Residential_Units", "Commercial_Units", "Total_Units"]
for col in unit_cols:
    df[col] = df[col].astype("Int64")


In [45]:
# Number of Rows
df.shape[0]

18491

In [46]:
# Number of Columns
df.shape[1]

21

We can see that there are 20 columns and 18491. The list shows that not all columns have the full amount of data. The amount of entries missing per column will be worked out and then if it is over 50%, the column will be dropped. If not then it will be interpolated.

Kuhn and Johnson (2013) say that if over 50% of values are missing the are often removed prior to analysis due to risk of introducing bias through imputation

In [47]:
# Finding out the percentage of missing values per column
def missing_values_percentage_table(df):

# Finding the number of missing values per column
    missing_values = df.isnull().sum()

# Finding the percentage of missing values by dividing it by the number of data entries there should be and multiplying by 100
    missing_value_perc =  (100 * missing_values / len(df))

# Concatenating the missing values and missing values %  along the top of the table
    missing_value_table = pd.concat([missing_values, missing_value_perc], axis=1)
    missing_value_table = missing_value_table.rename(columns={0: 'Missing Values', 1: '% of Total Values'})

# Sort the table by percentage of missing descending
    missing_values_table = missing_value_table.sort_values('% of Total Values', ascending=False)
    return missing_values_table

missing_percentage_df = missing_values_percentage_table(df)


display(missing_percentage_df.style.background_gradient(cmap='Blues'))

In [48]:
# Omitting Easement, Land_Square_Feet, Gross_Square_Feet, Commerical_Units, Apartment_number, Residential_Units

cols_to_drop = [
    "Easement",
    "Land_Square_Feet",
    "Gross_Square_Feet",
    "Commercial_Units",
    "Apartment_Number",
    "Residential_Units"
]

df = df.drop(columns=cols_to_drop)

Total Units - Interpolated through using the median Total Units of the neighbourhood

Year Built - Interpolated through using the modal Year Built of building category

In [49]:
# Interpolating Total Units by working out the median of the columns and filling the missing values with the median
df["Total_Units"] = df.groupby("Neighborhood")["Total_Units"]\
                      .transform(lambda x: x.fillna(x.median()))

In [50]:
# Interpolating Year Built by working out the modal value of the columns and filling the missing values with this modal value
df["Year_Built"] = df.groupby("Building_Class_Category")["Year_Built"]\
                     .transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else x.median()))


In [51]:
# Re-checking to see if missing values were filled accurately
missing_values = df.isnull().sum()
missing_values

Borough                           0
Neighborhood                      0
Building_Class_Category           0
Tax_Class_At_Present              0
Block                             0
Lot                               0
Building_Class_At_Present         0
Address                           0
Zip_Code                          0
Total_Units                       0
Year_Built                        0
Tax_Class_At_Time_Of_Sale         0
Building_Class_At_Time_Of_Sale    0
Sale_Price                        0
Sale_Date                         0
dtype: int64

We can see now that there is 0 missing values within the dataframe

In [52]:
# Within sale price column, some values show '0', this code removes their row from the dataframe
zeros = (df["Sale_Price"] == 0).sum()
print("Rows removed:", zeros)
df = df[df["Sale_Price"] != 0]


Rows removed: 4056


In [53]:
# Within the dataframe the neighbourhoods column is organised into their neighbourhoods, with some organised into even smaller sub sections.
# I have renamed the Neighbourhoods that fit this category so they are more generalised, meaning there is far less variation of neighbourhoods, allowing for a cleaner analysis.

neighborhood_organising = {
    # Greenwich Village
    "GREENWICH VILLAGE-CENTRAL": "GREENWICH VILLAGE",
    "GREENWICH VILLAGE-WEST": "GREENWICH VILLAGE",

    # Harlem
    "HARLEM-CENTRAL": "HARLEM",
    "HARLEM-EAST": "HARLEM",
    "HARLEM-UPPER": "HARLEM",
    "HARLEM-WEST": "HARLEM",

    # Midtown
    "MIDTOWN EAST": "MIDTOWN",
    "MIDTOWN WEST": "MIDTOWN",

    # Upper East Side
    "UPPER EAST SIDE (59-79)": "EAST SIDE",
    "UPPER EAST SIDE (79-96)": "EAST SIDE",
    "UPPER EAST SIDE (96-110)": "EAST SIDE",

    # Upper West Side
    "UPPER WEST SIDE (59-79)": "WEST SIDE",
    "UPPER WEST SIDE (79-96)": "WEST SIDE",
    "UPPER WEST SIDE (96-116)": "WEST SIDE",

    # Washington Heights
    "WASHINGTON HEIGHTS LOWER": "WASHINGTON HEIGHTS",
    "WASHINGTON HEIGHTS UPPER": "WASHINGTON HEIGHTS",

    #East Side
    "LOWER EAST SIDE": "EAST SIDE",
    "UPPER EAST SIDE": "EAST SIDE",
    #West Side
    "LOWER WEST SIDE": "WEST SIDE",
    "UPPER WEST SIDE": "WEST SIDE"
}

df["Neighborhood"] = df["Neighborhood"].replace(neighborhood_organising)


In [54]:
# Running the different Neighbourhood names to see if the previous code has worked
df["Neighborhood"].unique()

array(['ALPHABET CITY', 'CHELSEA', 'CHINATOWN', 'CIVIC CENTER', 'CLINTON',
       'EAST VILLAGE', 'FASHION', 'FINANCIAL', 'FLATIRON', 'GRAMERCY',
       'GREENWICH VILLAGE', 'HARLEM', 'INWOOD', 'JAVITS CENTER',
       'KIPS BAY', 'LITTLE ITALY', 'EAST SIDE', 'MANHATTAN VALLEY',
       'MIDTOWN CBD', 'MIDTOWN', 'MORNINGSIDE HEIGHTS', 'MURRAY HILL',
       'ROOSEVELT ISLAND', 'SOHO', 'SOUTHBRIDGE', 'TRIBECA', 'WEST SIDE',
       'WASHINGTON HEIGHTS'], dtype=object)

In [55]:
# Adding coordinates to each neighbourhood
neighborhood_coords = {
    "ALPHABET CITY": (40.7258, -73.9775),
    "CHELSEA": (40.7465, -74.00134),
    "CHINATOWN": (40.7130, -73.9960),
    "CIVIC CENTER": (40.7141, -74.0028),
    "CLINTON": (40.7500, -73.9934),
    "EAST VILLAGE": (40.7300, -73.9800),
    "FASHION": (40.7547, -73.9916),
    "FINANCIAL": (40.7030, -74.0123),
    "FLATIRON": (40.7401, -73.9903),
    "GRAMERCY": (40.7368, -73.9845),
    "GREENWICH VILLAGE": (40.7128, -74.0060),
    "HARLEM": (40.8116, -73.9465),
    "INWOOD": (40.8677, -73.9212),
    "JAVITS CENTER": (40.7569, -74.0017),
    "KIPS BAY": (40.7423, -73.9801),
    "LITTLE ITALY": (40.7191, -73.9973),
    "EAST SIDE": (40.7736, -73.9566),
    "MANHATTAN VALLEY": (40.7967, -73.9684),
    "MIDTOWN CBD": (40.7500, -73.9800),
    "MIDTOWN": (40.7549, -73.9840),
    "MORNINGSIDE HEIGHTS": (40.7128, -74.0060),
    "MURRAY HILL": (40.7479, -73.9757),
    "ROOSEVELT ISLAND": (40.7605, -73.9510),
    "SOHO": (40.7246, -74.0019),
    "SOUTHBRIDGE": (40.7287, -73.9922),
    "TRIBECA": (40.7163, -74.0086),
    "WEST SIDE": (40.7870, -73.9754),
    "WASHINGTON HEIGHTS": (40.8417, -73.9394)
}

# Converting dictionary to fromat within dataframe
coords_df = (
    pd.DataFrame.from_dict(
        neighborhood_coords,
        orient="index",
        columns=["Latitude", "Longitude"]
    )
    .reset_index()
    .rename(columns={"index": "Neighborhood"})
)
# Standardising the names of the Neighbourhoods
df["Neighborhood"] = df["Neighborhood"].str.strip().str.upper()
coords_df["Neighborhood"] = coords_df["Neighborhood"].str.strip().str.upper()

# Merging into main dataframe
df = df.merge(coords_df, on="Neighborhood", how="left")




In [56]:
# Making new csv with updated data entries
df_to_csv = df.to_csv('Manhattan_Property_Dataset_Updated.csv')

# **Data Analysis**

#### **Mean Sale Prices per Neighbourhood**

In [57]:
# Which neibourhood is the most expensive?

# Grouping neighbourhoods and calculating mean sale price
neighborhood_mean = (
    df.groupby("Neighborhood", as_index=False)["Sale_Price"]
      .mean()
      .rename(columns={"Sale_Price": "Mean_Sale_Price"})
)

# Sorting by mean sale price in descending order
neighborhood_mean = neighborhood_mean.sort_values(
    "Mean_Sale_Price", ascending=False
)

# Plot with Plotly Express
fig = px.bar(
    data_frame=neighborhood_mean, # Added data_frame argument
    x="Neighborhood",
    y="Mean_Sale_Price",
    title="Most Expensive Neighborhoods (Mean Sale Price)",
    labels={"Neighborhood": "Neighborhood", "Mean_Sale_Price": "Mean Sale Price ($)"},
)
fig.update_layout(
    xaxis_tickangle=-45
)
fig.show()

This graph clearly shows us which neighbourhoods are the most expensive. Based on mean sale price, it shows the average sale price within each neighbourhood in the from of an interactive bar chart.

By hovering over the bar, you can see the sale price and which neighbourhood it is.

We can see Midtown Central Business District (CBD) is the most expensive, followed by Fashion and Clinton and Soho. The rest are all very similarly priced with Inwood the cheapest neighbourhood to live in in Manhattan.

#### **Number of Sales per Neighbourhood**

In [76]:
# Aggregate sales per neighbourhood
sales_map = (
    df.groupby(["Neighborhood", "Latitude", "Longitude"], as_index=False)
      .size()
      .rename(columns={"size": "Number_of_Sales"})
)

fig = px.scatter_mapbox(
    sales_map,
    lat="Latitude",
    lon="Longitude",
    size="Number_of_Sales",
    size_max=40,
    color="Number_of_Sales",
    hover_name="Neighborhood",
    hover_data={
        "Number_of_Sales": True,
        "Latitude": False,
        "Longitude": False
    },
    zoom=11,
    mapbox_style="carto-positron",
    title="Sales Volume by Neighborhood"
)

fig.update_layout(
    height=1000,
    width=1800,
    margin=dict(l=20, r=20, t=60, b=20)
)

fig.show()




In [58]:
sales_per_neighborhood = (
    df.groupby("Neighborhood")
      .size()
      .reset_index(name="Number_of_Sales")
      .sort_values("Number_of_Sales", ascending=False)
)
fig = px.bar(
    sales_per_neighborhood,
    x="Number_of_Sales",
    y="Neighborhood",
    orientation="h",
    title="Number of Property Sales per Neighborhood"
)

fig.update_layout(
    height=1000,
    bargap=0.25,
    yaxis={'categoryorder':'total ascending'})
fig.show()

#### **Average Property Price per Neighbourhood over time**

In [60]:

# Ensure datetime + numeric
df["Sale_Date"] = pd.to_datetime(df["Sale_Date"], errors="coerce")
df["Sale_Price"] = pd.to_numeric(df["Sale_Price"], errors="coerce")

# Extract year and remove invalid rows
df["Sale_Year"] = df["Sale_Date"].dt.year
df = df[(df["Sale_Year"].notna()) & (df["Sale_Price"] > 0)]

# ===== 2) Aggregate: mean sale price per year per neighborhood =====
yearly_neigh_mean = (
    df.groupby(["Sale_Year", "Neighborhood"], as_index=False)["Sale_Price"]
      .mean()
)

# Optional: keep only neighborhoods with enough data points (uncomment if needed)
# valid_neigh = yearly_neigh_mean["Neighborhood"].value_counts()
# yearly_neigh_mean = yearly_neigh_mean[yearly_neigh_mean["Neighborhood"].isin(valid_neigh[valid_neigh >= 3].index)]

# ===== 3) Build interactive dropdown (individual neighborhood selection) =====
fig = go.Figure()

neighborhoods = sorted(yearly_neigh_mean["Neighborhood"].dropna().unique())

# Add one trace per neighborhood (start hidden)
for neigh in neighborhoods:
    df_n = yearly_neigh_mean[yearly_neigh_mean["Neighborhood"] == neigh]
    fig.add_trace(
        go.Scatter(
            x=df_n["Sale_Year"],
            y=df_n["Sale_Price"],
            mode="lines",
            name=neigh,
            visible=False
        )
    )

# Create dropdown buttons: one per neighborhood + "All"
buttons = []

# "All Neighborhoods"
buttons.append(
    dict(
        label="All Neighborhoods",
        method="update",
        args=[
            {"visible": [True] * len(neighborhoods)},
            {"title": "Mean Sale Price per Year – All Neighborhoods"}
        ],
    )
)

# Each neighborhood individually
for i, neigh in enumerate(neighborhoods):
    visible = [False] * len(neighborhoods)
    visible[i] = True
    buttons.append(
        dict(
            label=neigh,
            method="update",
            args=[
                {"visible": visible},
                {"title": f"Mean Sale Price per Year – {neigh}"}
            ],
        )
    )

# Default view: show all
for tr in fig.data:
    tr.visible = True

# ===== 4) Layout + show =====
fig.update_layout(
    updatemenus=[dict(
        type="dropdown",
        buttons=buttons,
        x=1.15,
        y=1,
        showactive=True
    )],
    title="Mean Sale Price per Year – All Neighborhoods",
    xaxis_title="Year",
    yaxis_title="Mean Sale Price ($)",
)

fig.show()


#### **Building Class Frequency**

In [62]:
df["Building_Class_Group"] = df["Building_Class_At_Present"].str[0]
bc_group_counts = (
    df["Building_Class_Group"]
      .value_counts()
      .reset_index()
)
bc_group_counts.columns = ["Building_Class_Group", "Count"]
bc_group_counts = bc_group_counts[bc_group_counts['Count'] >= 50]

fig = px.bar(
    bc_group_counts,
    x="Count",
    y="Building_Class_Group",
    orientation="h",
    title="Frequency of Building Class Groups",
    labels={
        "Count": "Number of Sales",
        "Building_Class_Group": "Building Class Group"
    }
)

fig.update_layout(
    height=800,
    bargap=0.15,
    yaxis={"categoryorder": "total ascending"}
)

fig.show()

#### **Average Sale Price per Building Class**

Building Class Types Key:

* R - Condominiums (Condos)
* D - Elevator Apartments
* C - Walk-Up Apartments
* O - Office Buildings
* A - Single Family Homes
* B - Two Family Homes
* K - Mixed Use Buildings
* S - Cooperative Apartments




In [63]:
import pandas as pd
import plotly.express as px

# 1️⃣ Create building class group (first letter)
df["Building_Class_Group"] = df["Building_Class_At_Present"].str[0]

# 2️⃣ Keep only building classes with more than 50 observations
class_counts = df["Building_Class_Group"].value_counts()
valid_classes = class_counts[class_counts > 50].index

df_filtered = df[df["Building_Class_Group"].isin(valid_classes)]

# 3️⃣ Create box plots using Plotly Express
fig = px.box(
    df_filtered,
    y="Building_Class_Group",
    x="Sale_Price",
    title="Sale Price Distribution by Building Class (Classes with > 50 Sales)",
    labels={
        "Building_Class_Group": "Building Class",
        "Sale_Price": "Sale Price ($)"
    }
)

# Strongly recommended for property prices
fig.update_xaxes(type="log")

fig.show()


#### **Year Built and Sale Price Relationship**

In [64]:


# Keep only plausible build years (adjust if needed)
df_plot = df[df["Year_Built"].between(1800, 2025)]

fig = px.scatter(
    df_plot,
    x="Year_Built",
    y="Sale_Price",
    trendline="ols",
    opacity=0.4,
    title="Relationship Between Sale Price and Year Built",
    labels={"Year_Built": "Year Built", "Sale_Price": "Sale Price ($)"}
)
fig.update_yaxes(type="log")
fig.update_traces(
    marker=dict(size=5),
    line=dict(width=1.5)
)


fig.show()


#### **Total Units and Sale Price Relationship**

In [65]:
df_plot = df[
    (df["Total_Units"] > 0) &
    (df["Sale_Price"] > 0)
]

fig = px.scatter(
    df_plot,
    x="Total_Units",
    y="Sale_Price",
    trendline="ols",
    opacity=1,
    title="Relationship Between Total Units and Sale Price",
    labels={
        "Total_Units": "Total Units",
        "Sale_Price": "Sale Price ($)"
    }
)

# Log–log scale (very important)
fig.update_xaxes(type="log")
fig.update_yaxes(type="log", tickprefix="$")

fig.update_layout(template="plotly_white")
fig.show()

#### **Price Per Unit**

In [66]:
import plotly.express as px

# Keep valid rows
df_units = df[
    (df["Total_Units"] > 0) &
    (df["Sale_Price"] > 0)
]

import plotly.express as px

# Keep valid rows
df_units = df[
    (df["Total_Units"] > 0) &
    (df["Sale_Price"] > 0)
]

# Limit to reasonable unit counts for exact comparison
df_units = df_units[df_units["Total_Units"] <= 50]

# Mean sale price per unit count
avg_price_by_units = (
    df_units
    .groupby("Total_Units", as_index=False)["Sale_Price"]
    .mean()
)

fig = px.line(
    avg_price_by_units,
    x="Total_Units",
    y="Sale_Price",
    markers=True,
    title="Average Sale Price by Number of Units",
    labels={
        "Total_Units": "Number of Units",
        "Sale_Price": "Average Sale Price ($)"
    }
)

fig.update_yaxes(type="log", tickprefix="$")
fig.update_layout(template="plotly_white")

fig.show()


# References

* Kuhn, M., & Johnson, K. (2013). Applied Predictive Modeling. Springer.